pkg/operator/controller/service-dns/controller_test.go

```diff
@@ -0,0 +1,267 @@
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: "openshift-ingress",
+				Name:      name,
+			},
+			Spec: iov1.DNSRecordSpec{
+				DNSName:             dnsName,
+				RecordType:          iov1.CNAMERecordType,
+				Targets:             targets,
+				RecordTTL:           30,
+				DNSManagementPolicy: iov1.ManagedDNS,
+			},
+		}
+	}
+	req := func(ns, name string) reconcile.Request {
+		return reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Namespace: ns,
+				Name:      name,
+			},
+		}
+	}
+	tests := []struct {
+		name             string
+		existingObjects  []runtime.Object
+		reconcileRequest reconcile.Request
+		expectCreate     []client.Object
+		expectUpdate     []client.Object
+	}{
+		{
+			name: "gateway with no listeners",
+			existingObjects: []runtime.Object{
+				gw("example-gateway"),
+				svc(
+					"example-gateway",
+					map[string]string{
+						"gateway.istio.io/managed": "example-gateway",
+					},
+					map[string]string{
+						"istio.io/gateway-name": "example-gateway",
+					},
+					ingHost("lb.example.com"),
+				),
+			},
+			reconcileRequest: req("openshift-ingress", "example-gateway"),
+			expectCreate:     []client.Object{},
+			expectUpdate:     []client.Object{},
+		},
+		{
+			name: "gateway with three listeners and two unique host names, no dnsrecords",
+			existingObjects: []runtime.Object{
+				gw(
+					"example-gateway",
+					l("stage-http", "*.stage.example.com", 80),
+					l("stage-https", "*.stage.example.com", 443),
+					l("prod-https", "*.prod.example.com", 443),
+				),
+				svc(
+					"example-gateway",
+					map[string]string{
+						"gateway.istio.io/managed": "example-gateway",
+					},
+					map[string]string{
+						"istio.io/gateway-name": "example-gateway",
+					},
+					ingHost("lb.example.com"),
+				),
+			},
+			reconcileRequest: req("openshift-ingress", "example-gateway"),
+			expectCreate: []client.Object{
+				dnsrecord("example-gateway-5bfc88bc87-wildcard", "*.prod.example.com", "lb.example.com"),
+				dnsrecord("example-gateway-57b76476b6-wildcard", "*.stage.example.com", "lb.example.com"),
+			},
+			expectUpdate: []client.Object{},
+		},
+		{
+			name: "gateway with two listeners and one dnsrecord with a stale target",
+			existingObjects: []runtime.Object{
+				gw(
+					"example-gateway",
+					l("http", "*.example.com", 80),
+					l("https", "*.example.com", 443),
+				),
+				svc(
+					"example-gateway",
+					map[string]string{
+						"gateway.istio.io/managed": "example-gateway",
+					},
+					map[string]string{
+						"istio.io/gateway-name": "example-gateway",
+					},
+					ingHost("newlb.example.com"),
+				),
+				dnsrecord("example-gateway-55bcfdb97d-wildcard", "*.example.com", "oldlb.example.com"),
+			},
+			reconcileRequest: req("openshift-ingress", "example-gateway"),
+			expectCreate:     []client.Object{},
+			expectUpdate: []client.Object{
+				dnsrecord("example-gateway-55bcfdb97d-wildcard", "*.example.com", "newlb.example.com"),
+			},
```

Comment:
What if a dnsRecord already exists with the same domain name as a gateway listener specifies? Will two dnsrecords be created? Is that worth a unit test?

Reply:
If it has the same host name, then it hashes to the same name for the CR, so the controller will update the existing CR rather than creating a new one.

Reply:
Or do you mean if two gateways specify the same host name?

Reply:
I meant if you have an outside actor (not the cluster-ingress-operator) creating dnsrecords, one with the same hostname as what you later create a Gateway Listener to have, what's the outcome? Now that I am saying that out loud, I remember you mentioning that admins are not supposed to create DNS Records by hand, so this is probably an unsupported situation.

Reply:
Yeah, the only thing that should ever create or update dnsrecords is cluster-ingress-operator itself.  (Nothing else should read dnsrecords either, other than must-gather.)